### PR TITLE
prov/efa: Adding CMocka Framework for unit testing

### PR DESCRIPTION
**Pull request by a-szegel**
Adding the CMocka framework for unit testing prov/efa.

The unit test I add calls efa_ah_alloc() twice with the same GID, and enforces that the function ibv_create_ah() only gets called once.

Signed-off-by: Seth Zegelstein <szegel@amazon.com>

### CODE REVIEW

File: prov/efa/test/README.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
+# EFA unit tests
+
+## How to run
+
+To run efa unit tests, you will need to have cmocka installed.
+See: https://cmocka.org/
```
Comment (mmubarak) [inline prov/efa/test/README.md:6]: Add install link for cmocka

File: prov/efa/test/README.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+To mock a function, you will need to add ```-Wl,--wrap=<function to mock>```
+to the Makefile.include as part of efatest_LIBS. Then, after declaring the function,
+you can replace it with __wrap_<function to mock>. If you need to use the original
+function, you can use __real_<function to mock> after declaring it.
+
+### Mocking static functions
+
+Static functions are not mockable due to the way they are compiled. Instead you
+will have to wrap a static function in a non static function. Currently we have
+no examples of this. If you would like to mock a static function, To make this
```
Comment (mmubarak) [inline prov/efa/test/README.md:34]: Why don't we have an example to mock a static function?

File: prov/efa/test/README.md [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+simpler, you can add a macro "unit_static" that will change static functions to
+non static when --enable-efa-unit-test is enabled. You will also need to add
```
Comment (mmubarak) [inline prov/efa/test/README.md:36]: We agreed on a different outcome from last week's meeting. Do we still want to mock static functions using 'unit_static' macro?

File: prov/efa/test/README.md [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+a new declaration by creating a file such as "efa_static_headers.h".
+
+### Manipulating mock behavior
+
+You can use the cmocka API to change the behavior of your mock function.
+
+The will_return class of functions will place values onto a stack that can
```
Comment (mmubarak) [inline prov/efa/test/README.md:43]: typo--> will return

File: prov/efa/test/efa_test_getinfo.c [hunk 5]
```diff
@@ -0,0 +1,32 @@
+#include "efa_unit_tests.h"
+
+struct ibv_device **__real_ibv_get_device_list(int *num_devices);
+
+/* This function is a bit tricky. Due to the fact
+ * that it is called during provider init, we can't
+ * use cmocka's mock_type and will_return. Thus this function
```
Comment (mmubarak) [inline prov/efa/test/efa_test_getinfo.c:7]: So we cannot mock functions that are called during provider init?

File: prov/efa/test/rdma_core_mocks.c [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+#include <stdint.h>
+#include <stdarg.h>
+#include <stddef.h>
+#include <setjmp.h>
+#include <cmocka.h>
+
+struct ibv_ah *__wrap_ibv_create_ah(struct ibv_pd *pd, struct ibv_ah_attr *attr) {
+    check_expected(pd);
+    check_expected(attr);
+    return (struct ibv_ah*) mock();
```
Comment (mmubarak) [inline prov/efa/test/rdma_core_mocks.c:14]: For clarification, can we add a comment with some details on what mock() is doing here?

File: prov/efa/test/README.md (hunk 3 above)
Comment (wckzhang) [reply prov/efa/test/README.md:36]: I didn't update the PR since that meeting

File: prov/efa/test/efa_test_getinfo.c (hunk 5 above)
Comment (wckzhang) [reply prov/efa/test/efa_test_getinfo.c:7]: We can. We just can't change their behavior per unit test. We can do it per binary though.

File: prov/efa/test/rdma_core_mocks.c (hunk 6 above)
Comment (a-szegel) [reply prov/efa/test/rdma_core_mocks.c:14]: Done

File: prov/efa/test/efa_test_getinfo.c (hunk 5 above)
Comment (a-szegel) [reply prov/efa/test/efa_test_getinfo.c:7]: We can mock them, but we need to hard code in their behavior.  Since they are called before the unit tests execute, the unit tests can't define their behavior.

File: prov/efa/test/README.md (hunk 3 above)
Comment (a-szegel) [reply prov/efa/test/README.md:36]: Removed.

File: prov/efa/test/README.md (hunk 2 above)
Comment (a-szegel) [reply prov/efa/test/README.md:34]: Mocking static functions is out of scope of the initial implementation of the unit testing framework.

File: prov/efa/test/README.md (hunk 1 above)
Comment (a-szegel) [reply prov/efa/test/README.md:6]: done

File: prov/efa/test/efa_unit_tests.c [hunk 7]
```diff
@@ -0,0 +1,14 @@
+#include "efa_unit_tests.h"
+
+int main(void)
+{
+	int ret;
+	const struct CMUnitTest efa_unit_tests[] = {
+		cmocka_unit_test(create_efa_ah_twice)    // Requires an EFA device to work
```
Comment (mjkoop) [inline prov/efa/test/efa_unit_tests.c:7]: Let's stick with the C-style comments :-)

Comment (a-szegel) [reply prov/efa/test/efa_unit_tests.c]: done :)

File: prov/efa/src/efa_av.c [hunk 8]
```diff
@@ -1075,3 +1075,97 @@ int efa_av_open(struct fid_domain *domain_fid, struct fi_av_attr *attr,
...
+};
+
+struct efa_resource *create_efa_resource() {
+    struct efa_resource *resource = (struct efa_resource*) malloc(sizeof(struct efa_resource));
+    resource->fabric = NULL;
+    resource->domain = NULL;
+    resource->av = NULL;
+
+    struct fi_info *info;
+    int rv = fi_getinfo(FI_VERSION(1, 14), NULL, NULL, 0ULL, NULL, &info);
```
Comment (wckzhang) [inline prov/efa/src/efa_av.c:1098]: We typically use "ret" for return value, would be a little more consistent to continue doing so

File: prov/efa/src/efa_av.c [hunk 9]
```diff
@@ -1075,3 +1075,97 @@ int efa_av_open(struct fid_domain *domain_fid, struct fi_av_attr *attr,
...
+
+    rv = fi_fabric(info->fabric_attr, &resource->fabric, NULL);
+    if (rv) {
+        free(resource);
+        return NULL;
+    }
+
+    rv = fi_domain(resource->fabric, info, &resource->domain, NULL);
+    if (rv) {
+        fi_close((fid_t) resource->fabric);
```
Comment (wckzhang) [inline prov/efa/src/efa_av.c:1112]: This seems somewhat out of the ordinary, typically we just point at the fid object like fi_close(&resource->fabric->fid);

File: prov/efa/src/efa_av.c [hunk 10]
```diff
@@ -1075,3 +1075,97 @@ int efa_av_open(struct fid_domain *domain_fid, struct fi_av_attr *attr,
...
+        free(resource);
+        return NULL;
+    }
+
+    struct fi_av_attr av_attr = {0};
```
Comment (wckzhang) [inline prov/efa/src/efa_av.c:1117]: Typically we put variable declarations at the top of the function

File: util/.gitignore [hunk 11]
```diff
@@ -0,0 +1 @@
+fi_info.trs
```
Comment (wckzhang) [inline util/.gitignore:1]: Should put this in a separate commit

File: prov/efa/test/efa_unit_tests.h [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+#include <stdarg.h>
+#include <stddef.h>
+#include <setjmp.h>
+#include <cmocka.h>
+#include "ofi_util.h"
+#include "stdio.h"
+#include "rxr.h"
+#include "include/ofi.h"
+
+/*  */
```
Comment (wckzhang) [inline prov/efa/test/efa_unit_tests.h:15]: Empty comment?

File: prov/efa/test/efa_unit_tests.h [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+
+#include <stdlib.h>
+#include <stdint.h>
+#include <stdarg.h>
+#include <stddef.h>
+#include <setjmp.h>
+#include <cmocka.h>
+#include "ofi_util.h"
+#include "stdio.h"
+#include "rxr.h"
```
Comment (wckzhang) [inline prov/efa/test/efa_unit_tests.h:12]: I'm pretty sure we don't need a lot of these includes right now. ofi_util, rxr, and ofi.h don't seem used

File: prov/efa/test/README.md [hunk 14]
```diff
@@ -0,0 +1,53 @@
+# EFA unit tests
+
+## How to run
+
+To run efa unit tests, you will need to have cmocka installed.
+See: https://github.com/clibs/cmocka/blob/master/INSTALL.md
+
+You will need to configure libfabric with ```--with-cmocka <path_to_cmocka_install>```.
```
Comment (wckzhang) [inline prov/efa/test/README.md:8]: Can you do ```--with-cmocka=<path_to_cmocka_install>``` to keep format consistent with example

File: prov/efa/test/README.md [hunk 15]
```diff
@@ -0,0 +1,53 @@
+# EFA unit tests
+
+## How to run
+
+To run efa unit tests, you will need to have cmocka installed.
+See: https://github.com/clibs/cmocka/blob/master/INSTALL.md
```
Comment (wckzhang) [inline prov/efa/test/README.md:6]: I'm not sure if you should be pointing to this repo - hasn't been updated since 2015 whereas the files on the website for cmocka.org have a version created in 2019

File: prov/efa/src/efa_av.c [hunk 16]
```diff
@@ -1075,3 +1075,97 @@ int efa_av_open(struct fid_domain *domain_fid, struct fi_av_attr *attr,
...
+    assert_non_null(efa_ah);
+
+    /* Call UUT again */
+    struct efa_ah *efa_ah2 = efa_ah_alloc(av, gid);
+    assert_true(efa_ah2 == efa_ah);
+    assert_true(efa_ah->refcnt == 2);
+    assert_true(av->ah_map->refcnt == 2);
+
+    delete_efa_resource(resource);
+    av = NULL;
```
Comment (wckzhang) [inline prov/efa/src/efa_av.c:1169]: This seems pointless as we're exiting the context of the function anyway

**Review by wckzhang — CHANGES_REQUESTED** [review]

File: prov/efa/src/efa_av.c (hunk 16 above)
Comment (a-szegel) [reply prov/efa/src/efa_av.c:1169]: done

File: prov/efa/test/README.md (hunk 15 above)
Comment (a-szegel) [reply prov/efa/test/README.md:6]: done

File: prov/efa/test/README.md (hunk 14 above)
Comment (a-szegel) [reply prov/efa/test/README.md:8]: done

File: prov/efa/test/efa_unit_tests.h (hunk 13 above)
Comment (a-szegel) [reply prov/efa/test/efa_unit_tests.h:12]: removed includes

File: prov/efa/test/efa_unit_tests.h (hunk 12 above)
Comment (a-szegel) [reply prov/efa/test/efa_unit_tests.h:15]: removed

File: util/.gitignore (hunk 11 above)
Comment (a-szegel) [reply util/.gitignore:1]: removed from MR.

File: prov/efa/src/efa_av.c (hunk 10 above)
Comment (a-szegel) [reply prov/efa/src/efa_av.c:1117]: done globally.

File: prov/efa/src/efa_av.c (hunk 9 above)
Comment (a-szegel) [reply prov/efa/src/efa_av.c:1112]: I was following an example in the NCCL Sim code Wei wrote.

File: prov/efa/src/efa_av.c (hunk 8 above)
Comment (a-szegel) [reply prov/efa/src/efa_av.c:1098]: changed rv to ret globally.

File: prov/efa/Makefile.include [hunk 17]
```diff
@@ -73,9 +73,26 @@ _efa_headers = \
...
+prov_efa_test_efa_unit_test_CPPFLAGS = -I$(top_srcdir)/include -pthread $(efatest_CPPFLAGS) -I$(top_srcdir)/prov/efa/src -I$(top_srcdir)/prov/efa/src/rxr -I$(top_srcdir)/prov/efa/test
+prov_efa_test_efa_unit_test_LDFLAGS = $(efatest_LDFLAGS) -Wl,--wrap=ibv_create_ah,--wrap=efadv_query_ah
+
+endif ENABLE_EFA_UNIT_TEST
+
 efa_CPPFLAGS += \
 	-I$(top_srcdir)/prov/efa/src/ \
-	-I$(top_srcdir)/prov/efa/src/rxr/
+	-I$(top_srcdir)/prov/efa/src/rxr/ \
+	-I$(top_srcdir)/prov/efa/test/
```
[the remark this answers is not in the document]
Comment (a-szegel) [reply prov/efa/Makefile.include:94]: Done

File: prov/efa/configure.m4 [hunk 18]
```diff
@@ -113,6 +115,37 @@ AC_DEFUN([FI_EFA_CONFIGURE],[
 	AC_SUBST(efa_CPPFLAGS)
 	AC_SUBST(efa_LDFLAGS)
 	AC_SUBST(efa_LIBS)
+
+    AC_ARG_WITH([cmocka],
```
[the remark this answers is not in the document]
Comment (a-szegel) [reply prov/efa/configure.m4:119]: I was using spaces, and I guess the entire file is tabs :(.

File: prov/efa/configure.m4 [hunk 19]
```diff
@@ -113,6 +115,37 @@ AC_DEFUN([FI_EFA_CONFIGURE],[
...
+	    [efa_unit_test=yes], [efa_unit_test=no])
+
+	if test x"$efa_unit_test" = x"yes"; then
+		AC_CHECK_HEADERS([cmocka.h],
+			[
+				AC_DEFINE([EFA_UNIT_TEST], [1],
+					[EFA unit testing])
+				efatest_CPPFLAGS="$efatest_CPPFLAGS $efadv_CPPFLAGS"
+				efa_CPPFLAGS="$efatest_CPPFLAGS $efadv_CPPFLAGS"
+				efatest_LIBS="-lcmocka -R$with_cmocka -libverbs -static $efatest_LIBS"
```
[the remark this answers is not in the document]
Comment (a-szegel) [reply prov/efa/configure.m4:132]: If you do: `--with-cmocka` and don't give an arg, you get -Ryes

File: prov/efa/configure.m4 (hunk 18 above)
[the remark this answers is not in the document]
Comment (wckzhang) [reply prov/efa/configure.m4:119]: All of libfabric is tabs

[the remark this answers is not in the document]
Comment (a-szegel) [reply prov/efa/configure.m4]: I guess I should uncheck the use spaces instead of tabs box from my editor then lol.

File: prov/efa/configure.m4 (hunk 19 above)
[the remark this answers is not in the document]
Comment (a-szegel) [reply prov/efa/configure.m4:132]: Fixed this behavior:
```
./configure  #works
./configure --with-cmocka #works
./configure --with-cmocka=/usr/local/lib64 #works
./configure --with-cmocka=/some/random/path # fails non-gracefully during build
```

File: prov/efa/configure.m4 [hunk 20]
```diff
@@ -113,6 +115,37 @@ AC_DEFUN([FI_EFA_CONFIGURE],[
...
+				efa_LIBS+=" -lcmocka -libverbs -static $efatest_LIBS"
```
[the remark this answers is not in the document]
Comment (a-szegel) [reply prov/efa/configure.m4:133]: Fixed this issue with the and check.

File: prov/efa/configure.m4 (hunk 19 above)
[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/configure.m4:132]: with-cmocka should point to cmocka root path, not cmocka/lib

[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/configure.m4:132]: There is a function `FI_CHECK_PACKAGE` you can use, which generate CPPFLAGS and LIBS for you

File: prov/efa/src/efa_av.c [hunk 21]
```diff
@@ -1075,3 +1075,102 @@ int efa_av_open(struct fid_domain *domain_fid, struct fi_av_attr *attr,
...
+	struct fid_fabric *fabric;
+	struct fid_domain *domain;
+	struct fid_av *av;
+};
+
+struct efa_resource *create_efa_resource() {
+    struct efa_resource *resource = (struct efa_resource*) malloc(sizeof(struct efa_resource));
+    struct fi_info *info;
+    struct fi_av_attr av_attr = {0};
+    int ret;
```
Comment (wzamazon) [inline prov/efa/src/efa_av.c:1095]: ret -> err. To be specific

File: prov/efa/src/efa_av.c (hunk 9 above)
Comment (wzamazon) [reply prov/efa/src/efa_av.c:1115]: William is right.

File: prov/efa/src/efa_av.c [hunk 22]
```diff
@@ -1075,3 +1075,102 @@ int efa_av_open(struct fid_domain *domain_fid, struct fi_av_attr *attr,
...
+    fi_close((fid_t) resource->fabric);
+    free(resource);
+}
+
+/*
+ * Only works on nodes with EFA devices
+ * This test calls efa_ah_alloc twice with the same GID,
+ * and verifies that ibv_create_ah only gets called once.
+ */
+void create_efa_ah_twice() {
```
Comment (wzamazon) [inline prov/efa/src/efa_av.c:1143]: The function (test case) name should be something like `test_duplicate_efa_ah_creation` to indicate it is a real test case.

File: prov/efa/src/efa_av.c [hunk 23]
```diff
@@ -1075,3 +1075,102 @@ int efa_av_open(struct fid_domain *domain_fid, struct fi_av_attr *attr,
...
+    struct efa_resource *resource = create_efa_resource();
+    struct efa_av *av;
+    int ibv_ret = 4242;
+    struct efa_ah *efa_ah;
+    struct efa_ah *efa_ah2;
+
+    assert_non_null(resource);
+    av = container_of(resource->av, struct efa_av, util_av.av_fid);
+    assert_non_null(av);
+    uint8_t gid[EFA_GID_LEN];
```
Comment (wzamazon) [inline prov/efa/src/efa_av.c:1154]: definition of `gid` should be at the beginning of the function.

**Review by wzamazon — CHANGES_REQUESTED** [review]

File: prov/efa/Makefile.include [hunk 24]
```diff
@@ -73,9 +73,26 @@ _efa_headers = \
...
+check_PROGRAMS += prov/efa/test/efa_unit_test
+TESTS += prov/efa/test/efa_unit_test
+nodist_prov_efa_test_efa_unit_test_SOURCES = \
+	prov/efa/test/efa_unit_tests.h \
+	prov/efa/test/efa_unit_tests.c \
+	prov/efa/test/rdma_core_mocks.c
+
+prov_efa_test_efa_unit_test_LDADD = $(efatest_LIBS) $(linkback)
+prov_efa_test_efa_unit_test_CPPFLAGS = -I$(top_srcdir)/include -pthread $(efatest_CPPFLAGS) -I$(top_srcdir)/prov/efa/src -I$(top_srcdir)/prov/efa/src/rxr -I$(top_srcdir)/prov/efa/test
+prov_efa_test_efa_unit_test_LDFLAGS = $(efatest_LDFLAGS) -Wl,--wrap=ibv_create_ah,--wrap=efadv_query_ah
```
[the remark this answers is not in the document]
Comment (a-szegel) [reply prov/efa/Makefile.include]: Done, needed -Wl,--wrap=some_fcn on every line

File: prov/efa/src/efa_av.c (hunk 23 above)
Comment (a-szegel) [reply prov/efa/src/efa_av.c:1154]: done

File: prov/efa/src/efa_av.c (hunk 22 above)
Comment (a-szegel) [reply prov/efa/src/efa_av.c:1143]: done

File: prov/efa/src/efa_av.c (hunk 21 above)
Comment (a-szegel) [reply prov/efa/src/efa_av.c:1095]: done

File: prov/efa/src/efa_av.c (hunk 9 above)
Comment (a-szegel) [reply prov/efa/src/efa_av.c:1115]: done

File: prov/efa/configure.m4 [hunk 25]
```diff
@@ -113,6 +115,37 @@ AC_DEFUN([FI_EFA_CONFIGURE],[
...
 	AC_SUBST(efa_LDFLAGS)
 	AC_SUBST(efa_LIBS)
+
+    AC_ARG_WITH([cmocka],
+	    [AS_HELP_STRING([--with-cmocka=DIR],
+			    [Provide path to where the Cmocka runtime libraries
+                 are installed.])],
+	    [efa_unit_test=yes], [efa_unit_test=no])
+
+	if test x"$efa_unit_test" = x"yes"; then
```
[the remark this answers is not in the document]
Comment (a-szegel) [reply prov/efa/configure.m4]: Works as intended now :).

File: prov/efa/configure.m4 (hunk 18 above)
[the remark this answers is not in the document]
Comment (a-szegel) [reply prov/efa/configure.m4]: Changed all my spaces to tabs.

File: prov/efa/configure.m4 (hunk 19 above)
[the remark this answers is not in the document]
Comment (a-szegel) [reply prov/efa/configure.m4]: Switched to using FI_CHECK_PACKAGE

File: prov/efa/configure.m4 [hunk 26]
```diff
@@ -113,6 +115,44 @@ AC_DEFUN([FI_EFA_CONFIGURE],[
...
+						 [_expect_any],
+						 [],
+						 [$with_cmocka],
+						 [$with_cmocka/lib64],
+						 [
+							AC_DEFINE([EFA_UNIT_TEST], [1], [EFA unit testing])
+							efatest_CPPFLAGS="$efatest_CPPFLAGS $efadv_CPPFLAGS -I$with_cmocka/include"
+							efa_CPPFLAGS="$efatest_CPPFLAGS $efadv_CPPFLAGS"
+							efa_LDFLAGS="-L$with_cmocka/lib64"
+							efatest_LDFLAGS="-L$with_cmocka/lib64"
```
Comment (wzamazon) [inline prov/efa/configure.m4:138]: As mentioned before, you do not need `efatest_CPPFLAGS` and `efatest_LDFLAGS`. You do not need to update `efa_CPPFLAGS` and `efa_LDFLAGS` either.

`FI_CHECK_PACKAGE` will find cmocka and set `cmocka_CPPFLAGS` and  `cmocka_LDFLAGS` for you.

you just use them in linking and compiling of `efa_test`

File: prov/efa/configure.m4 [hunk 27]
```diff
@@ -113,6 +115,44 @@ AC_DEFUN([FI_EFA_CONFIGURE],[
...
 	AC_SUBST(efa_LDFLAGS)
 	AC_SUBST(efa_LIBS)
+
+	AC_ARG_WITH([cmocka],
+		[AS_HELP_STRING([--with-cmocka=DIR],
+				[Provide path to where the Cmocka runtime libraries
+				 are installed.])],
+		[efa_unit_test=yes], [efa_unit_test=no])
+
+	if [test x"$efa_unit_test" = x"yes"] && [test x"$with_cmocka" != x"yes"]; then
```
Comment (wzamazon) [inline prov/efa/configure.m4:125]: should be 

```
if [test x"$with_efa_unit_test" != x"no" && test -n "$with_efa_unit_test"].
```

e.g. user specified the `--with-efa-unit-test` and the value is not `no`.

File: prov/efa/configure.m4 [hunk 28]
```diff
@@ -113,6 +115,44 @@ AC_DEFUN([FI_EFA_CONFIGURE],[
...
+		FI_CHECK_PACKAGE(cmocka,
+						 [cmocka.h],
+						 [cmocka],
+						 [_expect_any],
+						 [],
+						 [$with_cmocka],
+						 [$with_cmocka/lib64],
```
Comment (wzamazon) [inline prov/efa/configure.m4:132]: I do not think you need this, FI_CHECK_PACKAGE should find the correct libdir for you.

File: prov/efa/configure.m4 [hunk 29]
```diff
@@ -113,6 +115,44 @@ AC_DEFUN([FI_EFA_CONFIGURE],[
 	AC_SUBST(efa_CPPFLAGS)
 	AC_SUBST(efa_LDFLAGS)
 	AC_SUBST(efa_LIBS)
+
+	AC_ARG_WITH([cmocka],
```
Comment (wzamazon) [inline prov/efa/configure.m4:119]: It seems you are using `--with-cmocka` to indicate that efa unit test will be built. That is confusing. The command line argument should be `--with-efa-unit-test`, user can point cmocka directory to that. e.g. 

`--with-efa-unit-test=/path/to/cmocka`

File: prov/efa/configure.m4 [hunk 30]
```diff
@@ -113,6 +115,44 @@ AC_DEFUN([FI_EFA_CONFIGURE],[
...
+				 are installed.])],
+		[efa_unit_test=yes], [efa_unit_test=no])
+
+	if [test x"$efa_unit_test" = x"yes"] && [test x"$with_cmocka" != x"yes"]; then
+		FI_CHECK_PACKAGE(cmocka,
+						 [cmocka.h],
+						 [cmocka],
+						 [_expect_any],
+						 [],
+						 [$with_cmocka],
```
Comment (wzamazon) [inline prov/efa/configure.m4:131]: to properly handle the case that $with_cmocka is "yes", I think you will need anther variable `cmocka_dir`, and something like:
```
if [test x"$with_efa_unit_test" = x"yes"]; then
    cmocka_dir="";
else
   cmocka_dir="$with_efa_unit_test"
```

File: prov/efa/Makefile.include [hunk 31]
```diff
@@ -73,10 +73,30 @@ _efa_headers = \
...
+if ENABLE_EFA_UNIT_TEST
+check_PROGRAMS += prov/efa/test/efa_unit_test
+TESTS += prov/efa/test/efa_unit_test
+nodist_prov_efa_test_efa_unit_test_SOURCES = \
+	prov/efa/test/efa_unit_tests.h \
+	prov/efa/test/efa_unit_tests.c \
+	prov/efa/test/rdma_core_mocks.c
+
+prov_efa_test_efa_unit_test_LDADD = $(efatest_LIBS) $(linkback)
+prov_efa_test_efa_unit_test_CPPFLAGS = -I$(top_srcdir)/include -pthread $(efatest_CPPFLAGS) -I$(top_srcdir)/prov/efa/src -I$(top_srcdir)/prov/efa/src/rxr -I$(top_srcdir)/prov/efa/test
```
Comment (wzamazon) [inline prov/efa/Makefile.include:86]: It seems that we can use `$efa_CPPFLAGS $cmocka_CPPFLAGS`? we do not really need `efatest_CPPFLAGS`.

File: prov/efa/Makefile.include [hunk 32]
```diff
@@ -73,10 +73,30 @@ _efa_headers = \
...
+
+prov_efa_test_efa_unit_test_LDFLAGS = $(efatest_LDFLAGS) -Wl,--wrap=ibv_create_ah \
```
Comment (wzamazon) [inline prov/efa/Makefile.include:88]: similarly, we do not need `efatest_LDFAGS`, we can just use `$(efa_LDFLAGS) $(cmocka_LDFLAGS)`

File: prov/efa/configure.m4 [hunk 33]
```diff
@@ -113,6 +115,44 @@ AC_DEFUN([FI_EFA_CONFIGURE],[
 	AC_SUBST(efa_CPPFLAGS)
 	AC_SUBST(efa_LDFLAGS)
 	AC_SUBST(efa_LIBS)
+
+	AC_ARG_WITH([cmocka],
+		[AS_HELP_STRING([--with-cmocka=DIR],
+				[Provide path to where the Cmocka runtime libraries
```
Comment (wzamazon) [inline prov/efa/configure.m4:121]: not "Cmocka runtime library" any more.

File: prov/efa/configure.m4 [hunk 34]
```diff
@@ -113,6 +115,44 @@ AC_DEFUN([FI_EFA_CONFIGURE],[
...
+							AC_DEFINE([EFA_UNIT_TEST], [1], [EFA unit testing])
+							efatest_CPPFLAGS="$efatest_CPPFLAGS $efadv_CPPFLAGS -I$with_cmocka/include"
+							efa_CPPFLAGS="$efatest_CPPFLAGS $efadv_CPPFLAGS"
+							efa_LDFLAGS="-L$with_cmocka/lib64"
+							efatest_LDFLAGS="-L$with_cmocka/lib64"
+							efatest_LIBS=" -lcmocka -R$with_cmocka/lib64 -libverbs -static $efatest_LIBS"
+							efa_LIBS+=" -lcmocka -libverbs -static $efatest_LIBS"
+						 ],
+					 	 [
+							AC_MSG_WARN([Cannot compile EFA unit tests without cmocka.])
```
Comment (wzamazon) [inline prov/efa/configure.m4:143]: should be AC_MSG_ERROR here because user explicitly asked to compile efa_test.

File: prov/efa/src/efa_av.c [hunk 35]
```diff
@@ -1075,3 +1075,102 @@ int efa_av_open(struct fid_domain *domain_fid, struct fi_av_attr *attr,
 	free(av);
 	return ret;
 }
+
+#ifdef EFA_UNIT_TEST
```
Comment (wzamazon) [inline prov/efa/src/efa_av.c:1079]: As I mentioned before, define EFA_UNIT_TEST to 0 in the else branch and use `if` here instead of `ifdef`.

File: prov/efa/test/README.md [hunk 36]
```diff
@@ -0,0 +1,54 @@
+# EFA unit tests
+
+## How to run
+
+To run efa unit tests, you will need to have cmocka installed.
+Cmocka Mirror: https://cmocka.org/files/
+Install Instructions: https://gitlab.com/cmocka/cmocka/-/blob/master/INSTALL.md
+
+You will need to configure libfabric with ```--with-cmocka=<path_to_cmocka_install>```.
```
Comment (wzamazon) [inline prov/efa/test/README.md:9]: As I mentioned before, `with-cmocka -> with-efa-unit-test`

**Review by wzamazon — CHANGES_REQUESTED** [review]

File: prov/efa/src/efa_av.c (hunk 35 above)
Comment (a-szegel) [reply prov/efa/src/efa_av.c:1079]: done

File: prov/efa/configure.m4 (hunk 34 above)
Comment (a-szegel) [reply prov/efa/configure.m4:143]: done

File: prov/efa/test/README.md (hunk 36 above)
Comment (a-szegel) [reply prov/efa/test/README.md:9]: switching back to `--enable-efa-unit-test=DIR`

File: prov/efa/Makefile.include (hunk 31 above)
Comment (a-szegel) [reply prov/efa/Makefile.include:86]: removed efatest_CPPFLAGS

File: prov/efa/configure.m4 (hunk 29 above)
Comment (a-szegel) [reply prov/efa/configure.m4:119]: switched to `--enable-efa-unit-test=/path/to/cmocka`

File: prov/efa/configure.m4 (hunk 28 above)
Comment (a-szegel) [reply prov/efa/configure.m4]: done

File: prov/efa/configure.m4 (hunk 27 above)
Comment (a-szegel) [reply prov/efa/configure.m4]: updated to be accurate

File: prov/efa/configure.m4 (hunk 26 above)
Comment (a-szegel) [reply prov/efa/configure.m4]: removed.

File: prov/efa/Makefile.include [hunk 37]
```diff
@@ -73,10 +73,31 @@ _efa_headers = \
...
+	prov/efa/test/efa_unit_tests.c \
+	prov/efa/test/rdma_core_mocks.c
+
+prov_efa_test_efa_unit_test_LDADD = $(efatest_LIBS) $(linkback)
+prov_efa_test_efa_unit_test_CPPFLAGS = $(cmocka_CPPFLAGS) -I$(top_srcdir)/include -pthread -I$(top_srcdir)/prov/efa/src -I$(top_srcdir)/prov/efa/src/rxr -I$(top_srcdir)/prov/efa/test
+
+prov_efa_test_efa_unit_test_LDFLAGS = $(efa_LDFLAGS) $(cmocka_LDFLAGS) -Wl,--wrap=ibv_create_ah \
+																		-Wl,--wrap=efadv_query_ah
+
+efa_CPPFLAGS += -I$(top_srcdir)/prov/efa/test \
```
Comment (wzamazon) [inline prov/efa/Makefile.include:91]: If you move this to before `prov_efa_test_efa_unit_test_CPPFLAGS`, then `prov_efa_test_efa_unit_test_CPPFLAGS` will be able to use `efa_CPPFLAGS`.

File: prov/efa/configure.m4 [hunk 38]
```diff
@@ -113,6 +114,46 @@ AC_DEFUN([FI_EFA_CONFIGURE],[
...
+						 [cmocka.h],
+						 [cmocka],
+						 [_expect_any],
+						 [],
+						 [$cmocka_dir],
+						 [$cmocka_dir/lib64],
+						 [
+							AC_DEFINE([EFA_UNIT_TEST], [1], [EFA unit testing])
+							efatest_LIBS=" -lcmocka -libverbs -static $efatest_LIBS"
+							efa_LIBS+=" $cmocka_LDFLAGS -lcmocka -libverbs -static $efatest_LIBS"
```
Comment (wzamazon) [inline prov/efa/configure.m4:141]: I do not think you need to change `efatest_LIBS` and `efa_LIBS`.

File: prov/efa/configure.m4 [hunk 39]
```diff
@@ -113,6 +114,46 @@ AC_DEFUN([FI_EFA_CONFIGURE],[
...
+							AC_DEFINE([EFA_UNIT_TEST], [0], [EFA unit testing])
+							AC_MSG_ERROR([Cannot compile EFA unit tests without a valid Cmocka installation directory.])
+							efa_unit_test=no
+					 	 ],
+						 [
+							#include <stdarg.h>
+							#include <stddef.h>
+							#include <setjmp.h>
+					 	 ])
+	fi
```
Comment (wzamazon) [inline prov/efa/configure.m4:153]: Do not you need a `else` branch and to define `EFA_UNIT_TEST` to 0?

File: prov/efa/configure.m4 [hunk 40]
```diff
@@ -113,6 +114,46 @@ AC_DEFUN([FI_EFA_CONFIGURE],[
...
+
+	AC_ARG_ENABLE([efa-unit-test],
+		[AS_HELP_STRING([--enable-efa-unit-test=CMOCKA_INSTALL_DIR],
+				[Provide a path to the CMocka installation directory
+				 in order to enable EFA Unit Tests.])])
+
+	cmocka_dir=""
+	if [test x"$enable_efa_unit_test" = x"yes"]; then
+    	cmocka_dir=""
+	else
```
Comment (wzamazon) [inline prov/efa/configure.m4:126]: Can `enable_efa_unit_test` be `no`?

**Review by wzamazon — CHANGES_REQUESTED** [review]

File: prov/efa/configure.m4 (hunk 40 above)
Comment (a-szegel) [reply prov/efa/configure.m4:126]: enable_efa_unit_test can be no if you pass --disable-efa-unit-test, updated if check to handle it to:

```
if [ test x"$enable_efa_unit_test" != x"" && test x"$enable_efa_unit_test" != x"no" ];
```

File: prov/efa/configure.m4 (hunk 39 above)
Comment (a-szegel) [reply prov/efa/configure.m4:153]: It compiles without it, but I added it in for explicitness

File: prov/efa/configure.m4 (hunk 38 above)
Comment (a-szegel) [reply prov/efa/configure.m4]: got rid of efatest_LIBS, still need efa_LIBS b/c I can't seem to drag it into Makefile.include without everything breaking.  The most important flag in there is the -static.

**Review by wzamazon — APPROVED** [review]

File: prov/efa/test/rdma_core_mocks.c (hunk 6 above)
Comment (mmubarak) [reply prov/efa/test/rdma_core_mocks.c:26]: Thank-you

File: prov/efa/src/efa_av.c [hunk 41]
```diff
@@ -1075,3 +1075,102 @@ int efa_av_open(struct fid_domain *domain_fid, struct fi_av_attr *attr,
...
+}
+
+/*
+ * Only works on nodes with EFA devices
+ * This test calls efa_ah_alloc twice with the same GID,
+ * and verifies that ibv_create_ah only gets called once.
+ */
+void test_duplicate_efa_ah_creation() {
+	/* Setup Local Variables */
+	struct efa_resource *resource = create_efa_resource();
```
Comment (mmubarak) [inline prov/efa/src/efa_av.c:1145]: If create_efa_resource returns NULL because fi_get_info returns an error, should we handle that and not move forward in that case?

Comment (a-szegel) [reply prov/efa/src/efa_av.c]: I had an assert lower down `assert_non_null(resource);`. It got a little messy when I was shuffling definitions to the top of the function.  I took the separated declaration and assignment of resource to make this more clear.

**Review by mmubarak — APPROVED** [review]
